### PR TITLE
Honeypot for libssh CVE-2018-10933

### PR DESCRIPTION
**Pull request by dekelb**
Honeypot for libssh CVE-2018-10933

### CODE REVIEW

File: services/libssh/config.json [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+        },
+        {
+            "name": "ssh",
+            "label": "SSH Interaction",
+            "fields": ["originating_ip", "originating_port", "username", "password", "additional_fields"],
+            "policy": "Alert"
+        }
+    ],
+    "service": {
+        "allow_many": false,
```
Comment (lorg) [inline services/libssh/config.json]: Why not allow multiples on different ports?

File: services/libssh/config.json [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+        "supported_os_families": "Linux",
+        "ports": [
+        ],
+        "name": "cve_2018_10933",
```
Comment (lorg) [inline services/libssh/config.json]: Change the name and label to libssh

File: services/libssh/consts.py [hunk 3]
```diff
@@ -0,0 +1,19 @@
+# -*- coding: utf-8 -*-
+
+SERVER_SIG = "SSH-2.0-libssh_0.7.4"
+
+CVE_SSH_PORT = 2200
```
Comment (lorg) [inline services/libssh/consts.py]: Why is this default port different from the default value of the argument? If the port is not specified, which will be used?

File: services/libssh/cve_2018_10933_service.py [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+        super(CVEService, self).__init__(*args, **kwargs)
+        self.server = None
+        self.transport = None
+        self.chan = None
+
+    def alert(self, sock, event_type=CVE_ALERT_TYPE, *args, **kwargs):
+        """Send alert."""
+        params = {
+            EVENT_TYPE_FIELD_NAME: event_type,
+            ORIGINATING_IP_FIELD_NAME: sock.getpeername()[SOCK_IP_POSITION],
```
Comment (lorg) [inline services/libssh/cve_2018_10933_service.py]: Adding constants for 1 and 0 here is a bit over the top IMO.
You can also just do tuple unpacking e.g.
`ip, port = sock.getpeername()`
`params = ...`

File: services/libssh/cve_2018_10933_server.py [hunk 5]
```diff
@@ -0,0 +1,323 @@
...
+                self._expect_packet(MSG_KEXINIT)
+
+                while self.active:
+                    if self.packetizer.need_rekey() and not self.in_kex:
+                        self._send_kex_init()
+                    try:
+                        ptype, m = self.packetizer.read_message()
+                    except NeedRekeyException:
+                        continue
+                    if ptype == MSG_USERAUTH_SUCCESS:
```
Comment (lorg) [inline services/libssh/cve_2018_10933_server.py]: You should add a comment on the function header saying that that this code was copied from paramiko, and another comment for this line, saying that this if actually implements the detection of the exploit attempt of the vulnerability

File: services/libssh/cve_2018_10933_server.py [hunk 6]
```diff
@@ -0,0 +1,323 @@
...
+    transport = None
+    paramiko_server = None
+
+    def handle(self):
+        self.transport = CVETransport(self.connection, gss_kex=True)
+        self.transport.alert = self.alert
+        self.transport.set_gss_host(socket.getfqdn(""))
+        try:
+            self.transport.load_server_moduli()
+        except:
```
Comment (lorg) [inline services/libssh/cve_2018_10933_server.py]: Why add except: raise? I realize you probably used this for debugging, but do you still need this code here?

File: services/libssh/cve_2018_10933_server.py [hunk 7]
```diff
@@ -0,0 +1,323 @@
...
+
+
+class SSHServer(object):
+    def run(self, port):
+        requestHandler = SSHRequestHandler
+        requestHandler.alert = self.alert
+
+        self.server = ThreadingTCPServer(("", port), requestHandler)
+        self.server.socket.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
+        self.server.socket.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
```
Comment (lorg) [inline services/libssh/cve_2018_10933_server.py]: Why is this line repeated?

File: services/libssh/cve_2018_10933_server.py [hunk 8]
```diff
@@ -0,0 +1,323 @@
...
+        self.chan.close()
+
+
+class SSHServer(object):
+    def run(self, port):
+        requestHandler = SSHRequestHandler
+        requestHandler.alert = self.alert
+
+        self.server = ThreadingTCPServer(("", port), requestHandler)
+        self.server.socket.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
```
Comment (lorg) [inline services/libssh/cve_2018_10933_server.py]: You should add a comment about what is the purpose of these setsockopt calls.

File: services/libssh/config.json [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+            "policy": "Alert"
+        },
+        {
+            "name": "ssh",
+            "label": "SSH Interaction",
+            "fields": ["originating_ip", "originating_port", "username", "password", "additional_fields"],
+            "policy": "Alert"
+        }
+    ],
+    "service": {
```
Comment (lorg) [inline services/libssh/config.json]: Add a description for this honeypot and specify the detected vulnerability there. As an example look at the description for the micros honeypot.

File: services/libssh/consts.py (hunk 3 above)
Comment (dekelb) [reply services/libssh/consts.py]: Actually the usage of this const is on to run the server not through honeycomb, but directly (check main func).
I've changed this back to 2222.

File: services/libssh/requirements.txt [hunk 10]
```diff
@@ -0,0 +1,2 @@
+paramiko==2.4.2
```
Comment (omercnet) [inline services/libssh/requirements.txt]:
```suggestion
paramiko~=2.4.0
```

File: services/libssh/requirements.txt [hunk 11]
```diff
@@ -0,0 +1,2 @@
...
+python-gssapi==0.6.4
```
Comment (omercnet) [inline services/libssh/requirements.txt]:
```suggestion
python-gssapi~=0.6.0
```

File: services/libssh/libssh_service.py [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+    def on_server_start(self):
+        """Initialize service."""
+        port = int(self.service_args.get(CVE_PORT_FIELD))
+        self.server = SSHServer()
+        self.server.alert = self.alert
+        self.signal_ready()
+        self.server.run(port)
+
+    def test(self):
+        """Test function, still need to implement this."""
```
Comment (omercnet) [inline services/libssh/libssh_service.py]:
```suggestion
        """Test the service by conneting and passing the USERAUTH_SUCCESS msg"""
```

File: services/libssh/requirements.txt [hunk 13]
```diff
@@ -0,0 +1 @@
+paramiko==2.4.*
```
Comment (omercnet) [inline services/libssh/requirements.txt]:
```suggestion
paramiko~=2.4.0
```

File: services/libssh/cve_2018_10933_server.py [hunk 14]
```diff
@@ -0,0 +1,333 @@
...
+        return True
+
+    def check_channel_pty_request(
+        self, channel, term, width, height, pixelwidth, pixelheight, modes
+    ):  # noqa: D102
+        return True
+
+
+class SSHRequestHandler(StreamRequestHandler):  # noqa: D101
+
```
Comment (omercnet) [inline services/libssh/cve_2018_10933_server.py]:
```suggestion
    """SSH Request Handler."""

```

File: services/libssh/cve_2018_10933_server.py [hunk 15]
```diff
@@ -0,0 +1,333 @@
...
+        self.event.set()
+        return True
+
+    def check_channel_pty_request(
+        self, channel, term, width, height, pixelwidth, pixelheight, modes
+    ):  # noqa: D102
+        return True
+
+
+class SSHRequestHandler(StreamRequestHandler):  # noqa: D101
```
Comment (omercnet) [inline services/libssh/cve_2018_10933_server.py]:
```suggestion
class SSHRequestHandler(StreamRequestHandler):
```

File: services/libssh/cve_2018_10933_server.py [hunk 16]
```diff
@@ -0,0 +1,333 @@
...
+                KEY_FIELD_NAME: key.get_base64()
+            }
+        }
+        self.alert(self.socket, **data)
+        return paramiko.AUTH_FAILED
+
+    def get_allowed_auths(self, username):  # noqa: D102
+        return "password,publickey"
+
+    def check_channel_shell_request(self, channel):  # noqa: D102
```
Comment (omercnet) [inline services/libssh/cve_2018_10933_server.py]: please provide short description, this is for documentation purposes, excluding D102 is not accepted in this project :)

File: services/libssh/cve_2018_10933_server.py [hunk 17]
```diff
@@ -0,0 +1,338 @@
...
+# setup logging
+paramiko.util.log_to_file("demo_server.log")
+
+# Get current rsa key, or generate new one if needed
+key_file_path = os.path.join(os.path.dirname(__file__), "libssh_server_rsa.key")
+if not os.path.exists(key_file_path):
+    host_key = paramiko.RSAKey.generate(1024)
+    try:
+        host_key.write_private_key_file(key_file_path)
+    except:  # noqa: E722
```
Comment (omercnet) [inline services/libssh/cve_2018_10933_server.py]: this is new code, no reason not to catch an Exception like a normal human being..
```suggestion
    except Exception:
```

**Pull request merged**